### ISSUE TITLE
[Speed] parallel do stream wait to finish the copy operation

### ISSUE THREAD

**Issue by dzhwinter** [PaddlePaddle/Paddle#9012]
Currently, in our `parallel.do` we need to wait on streams to make sure that data has been distributed to GPU cards. Now our strategy is a for loop do the memory copy between cards. 
We all know that there is a primitive `scatter` in MPI http://mpitutorial.com/tutorials/mpi-scatter-gather-and-allgather/ , can distribute the data in a more efficient way. Will figure out the same way in NCCL.

**Comment by dzhwinter** [comment]
The NCCL library only have the `reduceScatter` operation and do not have the `scatter` primitive operation yet. 
https://github.com/NVIDIA/nccl/issues/67
So this would be a low priority thing since there would not be so much value on performance enhancement.

**Comment by dzhwinter** [comment]
This issue is a proposal, which has been implemented in the latest develop branch.
